File: old/2020/12_rain_risk/aoc12.jl
# Rain Risk
#
# Advent of Code 2020, day 12
# Solution by Geir Arne Hjelle, 2020-12-12

module AOC12

using Pipe


Directions = Dict('N' => [1, 0], 'E' => [0, 1], 'S' => [-1, 0], 'W' => [0, -1])
Bearings = Dict(0 => 'N', 90 => 'E', 180 => 'S', 270 => 'W')
Rotations = Dict(
    90 => (n, e) -> (-e, n),
    180 => (n, e) -> (-n, -e),
    270 => (n, e) -> (e, -n),
)

struct Position
    northing::Int32
    easting::Int32
    bearing::Int16
    waypoint_north::Int32
    waypoint_east::Int32

    # Initialize position without waypoint
    function Position(northing::Number, easting::Number, bearing::Number)
        new(northing, easting, (bearing + 360) % 360, 0, 0)
    end

    # Initialize position with waypoint
    function Position(northing::Number, easting::Number, bearing::Number, wpn::Number, wpe::Number)
        new(northing, easting, (bearing + 360) % 360, wpn, wpe)
    end
    
    # Initialize position relative to previous
    function Position(position::Position, dn::Number, de::Number, db::Number)
        new(
            position.northing + dn,
            position.easting + de,
            (position.bearing + db + 360) % 360,
            position.waypoint_north,
            position.waypoint_east,
        )
    end

    # Initialize position with waypoint by rotating waypoint
    function Position(position::Position, dn::Number, de::Number, rotation)
        wpn, wpe = rotation(position.waypoint_north, position.waypoint_east)
        new(
            position.northing + dn,
            position.easting + de,
            position.bearing,
            wpn,
            wpe,
        )
    end

    # Initialize position with waypoint relative to previous position
    function Position(position::Position, dn::Number, de::Number, dwpn::Number, dwpe::Number)
        new(
            position.northing + dn,
            position.easting + de,
            position.bearing,
            position.waypoint_north + dwpn,
            position.waypoint_east + dwpe,
        )
    end
end

struct Instruction
    action::Char
    value::Int32

    # Parse instruction from a line of text input
    function Instruction(instruction::String)
        new(instruction[1], @pipe instruction[2:end] |> parse(Int32, _))
    end
end


function move_ship(position, instruction)
    if instruction.action == 'R'
        Position(position, 0, 0, instruction.value)
    elseif instruction.action == 'L'
        Position(position, 0, 0, -instruction.value)
    elseif instruction.action == 'F'
        Position(
            position,
            (Directions[Bearings[position.bearing]] * instruction.value)...,
            0,
        )
    else
        Position(
            position,
            (Directions[instruction.action] * instruction.value)...,
            0,
        )
    end
end


function move_waypoint(position, instruction)
    if instruction.action == 'R'
        Position(position, 0, 0, Rotations[instruction.value])
    elseif instruction.action == 'L'
        Position(position, 0, 0, Rotations[360 - instruction.value])
    elseif instruction.action == 'F'
        Position(
            position,
            ([position.waypoint_north, position.waypoint_east] * instruction.value)...,
            0,
        )
    else
        Position(
            position,
            [0, 0]...,
            (Directions[instruction.action] * instruction.value)...,
        )
    end
end


function manhattan_distance(position)
    [position.northing, position.easting] .|> abs |> sum
end


# Solve the problem for one file
function solve(input)
    # Parse input
    instructions = split(input, "\n") .|> string .|> Instruction

    # Part 1
    part_1 = @pipe (
        instructions
        |> foldl(move_ship, _, init=Position(0, 0, 90))
        |> manhattan_distance
    )

    # Part 2
    part_2 = @pipe (
        instructions
        |> foldl(move_waypoint, _, init=Position(0, 0, 90, 1, 10))
        |> manhattan_distance
    )

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module

File: old/2020/12_rain_risk/aoc12.py
"""Rain Risk

Advent of Code 2020, day 12
Solution by Geir Arne Hjelle, 2020-12-16
"""
# Standard library imports
import enum
import pathlib
import sys
from dataclasses import dataclass, field
from typing import Callable, ClassVar, NamedTuple

# Third party imports
import parse

debug = print if "--debug" in sys.argv else lambda *_: None

_INSTRUCTION = parse.compile("{action:1}{value:d}")


class Action(NamedTuple):
    method: Callable
    params: list

    def __repr__(self):
        return f"{self.method}({', '.join(str(p) for p in self.params)})"


class Instruction(NamedTuple):
    action: Action
    value: int


class Bearing(enum.IntEnum):
    NORTH = 0
    EAST = 90
    SOUTH = 180
    WEST = 270

    @property
    def char(self):
        """Get character representing bearing"""
        return self.name[0]

    def __add__(self, dvalue):
        value = (self.value + dvalue) % 360
        return getattr(
            self, {v: n for n, v in self.__class__.__members__.items()}[value]
        )


class Position(NamedTuple):
    lat: int
    lon: int
    bearing: Bearing = Bearing.EAST


@dataclass
class Ship:
    position: Position = field(init=False)
    instructions: list = field(default_factory=list)
    actions: ClassVar[dict] = {}

    def __post_init__(self):
        """Initialize position"""
        self.compiled = [
            Instruction(self.actions[n["action"]], n["value"])
            for i in self.instructions
            for n in [_INSTRUCTION.parse(i)]
        ]
        self.position = Position(0, 0, Bearing.EAST)

    def reset(self):
        """Reset position"""
        self.__post_init__()

    def sail(self):
        """Sail according to all instructions"""
        self.reset()
        for instruction in self.compiled:
            self.execute(instruction)
        print(self)

    def execute(self, instruction, log=True):
        """Execute one instruction"""
        action, value = instruction
        getattr(self, action.method)(value, *action.params)
        if log:
            debug(f"{str(instruction):<50} {self.position}")

    def distance(self, to=Position(0, 0)):
        """Calculate Manhattan distance to other position"""
        return sum(abs(p - t) for p, t in zip(self.position[:2], to[:2]))

    def __str__(self):
        pos = self.position
        return (
            f"The ship is at ({pos.lat}, {pos.lon}), "
            f"distance to origin: {self.distance()}"
        )


class ShipModel1(Ship):
    actions = {
        "N": Action("move", (1, 0)),
        "E": Action("move", (0, 1)),
        "S": Action("move", (-1, 0)),
        "W": Action("move", (0, -1)),
        "L": Action("turn", (-1,)),
        "R": Action("turn", (1,)),
        "F": Action("forward", ()),
    }

    def move(self, multiplier, dlat, dlon):
        """Move the ship, keep the bearing"""
        pos = self.position
        self.position = Position(
            pos.lat + dlat * multiplier, pos.lon + dlon * multiplier, pos.bearing
        )

    def turn(self, multiplier, dbearing):
        """Turn the ship, keep the position"""
        pos = self.position
        self.position = Position(pos.lat, pos.lon, pos.bearing + dbearing * multiplier)

    def forward(self, steps):
        """Move the ship forward"""
        self.execute(
            Instruction(self.actions[self.position.bearing.char], steps), log=False
        )


class ShipModel2(Ship):
    actions = {
        "N": Action("move_waypoint", (1, 0)),
        "E": Action("move_waypoint", (0, 1)),
        "S": Action("move_waypoint", (-1, 0)),
        "W": Action("move_waypoint", (0, -1)),
        "L": Action("rotate_waypoint", (-1,)),
        "R": Action("rotate_waypoint", (1,)),
        "F": Action("forward", ()),
    }
    rotations = {
        90: (lambda lat, lon: (-lon, lat)),
        180: (lambda lat, lon: (-lat, -lon)),
        270: (lambda lat, lon: (lon, -lat)),
    }

    def __post_init__(self):
        super().__post_init__()
        self.waypoint = Position(1, 10)

    def move_waypoint(self, multiplier, dlat, dlon):
        """Move the waypoint"""
        wpt = self.waypoint
        self.waypoint = Position(
            wpt.lat + dlat * multiplier, wpt.lon + dlon * multiplier
        )

    def rotate_waypoint(self, multiplier, dangle):
        """Rotate the waypoint"""
        wpt = self.waypoint
        self.waypoint = Position(*self.rotations[(dangle * multiplier) % 360](*wpt[:2]))

    def forward(self, steps):
        """Move the ship towards the waypoint"""
        pos, wpt = self.position, self.waypoint
        self.position = Position(pos.lat + wpt.lat * steps, pos.lon + wpt.lon * steps)


def main(args):
    """Solve the problem for all file paths"""
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        solve(file_path)


def solve(file_path):
    """Solve the problem for one file path"""
    print(f"\n{file_path}:")
    instructions = file_path.read_text().strip().split("\n")

    # Part 1
    ship = ShipModel1(instructions)
    ship.sail()

    # Part 2
    ship = ShipModel2(instructions)
    ship.waypoint = Position(1, 10)
    ship.sail()


if __name__ == "__main__":
    main(sys.argv[1:])
